pkg/vms/hostvmslist.es6

```diff
@@ -0,0 +1,116 @@
+        </span>
+      </div>
+    </div>
+  )
+}
+
+// TODO: Add actual usage
+function VmUsageCharts ({vm}) {
+  return (<div>TBD: Usage Charts</div>);
+}
+
+// TODO: render allowed actions only (based on vm.state)
+function VmActions ({vm, onReboot, onForceReboot, onShutdown, onForceoff}) {
+  return (
+    <div className="dropdown pull-right dropdown-kebab-pf">
+      <button className="btn btn-link dropdown-toggle" type="button" id="dropdownKebabRight" data-toggle="dropdown" aria-haspopup="true" aria-expanded="true">
+        <span className="fa fa-ellipsis-v"></span>
+      </button>
+      <ul className="dropdown-menu dropdown-menu-right" aria-labelledby="dropdownKebabRight">
+        <li onClick={onReboot}><a href="#"><i className="fa fa-refresh"/> Reboot</a></li>
+        <li onClick={onForceReboot}><a href="#"><i className="fa fa-refresh"/> Force Reboot</a></li>
+        <li onClick={onShutdown}><a href="#"><i className="fa fa-power-off"/> Shutdown</a></li>
+        <li onClick={onForceoff}><a href="#"><i className="fa fa-power-off"/> Force Off</a></li>
+      </ul>
+    </div>
+);
+}
+
+function Vm ({ vm, onVmClick, onShutdown, onForceoff, onReboot, onForceReboot, onStart }) {
+  return (
+    <div className="list-group-item list-view-pf-stacked">
+      <div className="list-view-pf-actions">
+        <VmActions vm={vm} onReboot={onReboot} onForceReboot={onForceReboot} onShutdown={onShutdown} onForceoff={onForceoff} />
+      </div>
+
+      <div className="list-view-pf-main-info">
+        <div className="list-view-pf-left">
+          <StateIcon state={vm.state} onStart={onStart} />
+        </div>
+
+        <div className="list-view-pf-body">
+          <div className="list-view-pf-description">
+            <div className="list-group-item-heading">
+              <a onClick={onVmClick}>{vm.name}</a>
+              <small>Memory: {vm.currentMemory}</small>
+              <small>CPUs: {vm.vcpus}</small>
+            </div>
+          </div>
+
+          <div className="list-view-pf-additional-info">
+            <VmUsageCharts vm={vm} />
+          </div>
+        </div>
+      </div>
+    </div>
+)}
+
+function HostVmsList({ vms, dispatch }) {
+  let rows = [];
+  if (vms.length === 0) {
+    rows = <NoVm />;
+  } else {
+    rows = vms.map(vm => (
+      <Vm vm={vm}
+        onVmClick={() => dispatch(navigate(['vm', vm.name], true))}
+        onReboot={() => dispatch(rebootVm(vm.name))}
+        onForceReboot={() => dispatch(forceRebootVm(vm.name))}
+        onShutdown={() => dispatch(shutdownVm(vm.name))}
+        onForceoff={() => dispatch(forceVmOff(vm.name))}
+        onStart={() => dispatch(startVm(vm.name))}/>
+    ));
+  }
+
+  return (
+    <div className="container-fluid">
+      <div className="panel panel-default">
+        <div className="panel-heading">
+          Running Virtual Machines
+          <div className="actions pull-right">
+            <ReloadSwitch
```

Comment:
That's what ract/redux does. If the redux store is updated, the subscribed listener (means react virtual dom) is refreshed. Then React compares with the actual DOM and makes changes.
With redux there is (or should not be) any reason to care about this kind of optimization.

For completeness, the Refresh switch turns on/off the polling. The polling needs to be on to read usage details (historical data will be collected by JavaScript) and get notified of changes.

Polling is not nice in general but I don't see any other option now.

The very first implementation was based on machined but I dropped that code.
Machined signals new/deleted VMs but the provided VM properties set is very limited. Since we still need to poll libvirt to get actual VM usage, these 2 signals are not needed.

It's design decision that this Refresh Switch component is rendered as part of the HostVmsList. There's no need for that, it just seems to be compact.